Fixes for running cloudserver on s3c

francoisferrand opened the pull request:
- fix project path
- fix utapi config
- support null compatibility mode
- Enable multiObjectDelete optim' for mongo only
- do not fail on head with part number if object is empty
- All headObject on part1 for non mpu objects
- Allow disabling versioning if replication rules are disabled
- Remove useless storage class update
- Allow configuring valid storage classes
- Support batchDelete route when bucketName is not present
- Delete data when replacing object metadata with an empty object
- Trigger replication on putACL only for CRR

Issue: CLDSRV-638

Commit:
fix project path

Commit:
fix utapi config

Commit:
Allow disabling versioning if replication rules are disabled

Issue: CLDSRV-638

Commit:
Trigger replication on putACL only for CRR

When ACL are updated, check the updated replication info: if replicating
to cloud locations (i.e. zenko's multi-backend replication), then we
do not update replication info, to ensure ACL are not replicated.

However, if we are replicating to CRR location, then we should replicate
ACL, and we will now reset the replicationInfo.

Issue: CLDSRV-638

francoisferrand committed:
support null compatibility mode

null version compatibility mode should be enforced when using mongo.

Issue: CLDSRV-638

francoisferrand committed:
Enable multiObjectDelete optim' for mongo only

Issue: CLDSRV-638

francoisferrand committed:
Remove useless storage class update

Code tried to set storage-class from bucket's default location when
cold, but
- a bucket cannot have a cold location as default;
- storage location is per-object, not per bucket;
- it is set as expected in `collectResponseHeaders()`, in the next line.

Take the chance to create proper "mocks" for the different "states" of
archived/cold objects. Eventually this should be moved to Arsenal,
along with models, to create a corpus of possible object states and
enforce valid transitions in the model APIs.

Issue: CLDSRV-638

francoisferrand committed:
Allow configuring valid storage classes

The only supported storage class and should be STANDARD. We have no use
for other storage classes (in requests), and should not try to accept
anything else as it may affect the understanding apps have of their
use of the storage.

However, to ease transition, allow overriding the list of "valid"
storage classes by specifying the `VALID_STORAGE_CLASSES` environment
variable. This is a comma-separate list of supported values.

Practically, the parameter is not very useful:
* It is _validated_ in putObject, but not otherwise used: if storage
 class is specified, it has not effect on metadata or storage;
* It is added to the object when calling `copyObject`, with no further
  validation. To match behavior of 7.x branches, all used values should
  be added to the env variable;
* In initiate MPU, the storage class is stored in the MPU overview doc
  only it is also one of 'standard', 'standard_ia', 'reduced_redundancy'
  (in lowercase!).

By setting `VALID_STORAGE_CLASSES` appropriately, it is possible to get
the same behavior as 7.x branch; and this will be cleaned'up in 9.1
branch through the CLDSRV-639 ticket.

Issue: CLDSRV-638

francoisferrand committed:
Support batchDelete route when bucketName is not present

This parameter was not originally present, and is not enforced at the
moment (even though it seems to be set on every "production" call from
backbeat). At the moment, this parameter is mostly used only for quotas'
inflight update.

This should be cleaned' in CLDSRV-643, to settle on a proper API.
However, gracefully handle the case where the field is not set, so that:
- We ensure no regression in S3C when upgrading from 7 to 9;
- We give a change to multibackend's routeBackbeat tests to pass
  (CLDSRV-394) as they are still using the old route without bucket.

Issue: CLDSRV-638

francoisferrand committed:
Delete data when replacing object metadata with an empty object

Avoid creating orphaned data when overwriting an object with an empty object
using backbeat putMetadata route. This fixes a regression from CLDSRV-241,
which introduced the ability to update metadata (esp. locations) wihtout
deleting the data, used in case of cold storage: as data management is
handle by the garbagecollector service.

Issue: CLDSRV-638

bert-e commented:
# Hello francoisferrand,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://bert-e.scality.net/scality/cloudserver/doc/user).


<details>
  <summary><b>Available options</b></summary>

  name   | description  | privileged   | authored
  ------ | ------------ | ------------ |----------
  `/after_pull_request` | Wait for the given pull request id to be merged before continuing with the current one. |  | 
  `/bypass_author_approval` | Bypass the pull request author's approval |  :star:  | 
  `/bypass_build_status` | Bypass the build and test status |  :star:  | 
  `/bypass_commit_size` | Bypass the check on the size of the changeset ```TBA``` |  :star:  | 
  `/bypass_incompatible_branch` | Bypass the check on the source branch prefix |  :star:  | 
  `/bypass_jira_check` | Bypass the Jira issue check |  :star:  | 
  `/bypass_peer_approval` | Bypass the pull request peers' approval |  :star:  | 
  `/bypass_leader_approval` | Bypass the pull request leaders' approval |  :star:  | 
  `/approve` | Instruct Bert-E that the author has approved the pull request. |  |  :writing_hand: 
  `/create_pull_requests` | Allow the creation of integration pull requests. |  | 
  `/create_integration_branches` | Allow the creation of integration branches. |  | 
  `/no_octopus` | Prevent Wall-E from doing any octopus merge and use multiple consecutive merge instead |  | 
  `/unanimity` | Change review acceptance criteria from `one reviewer at least` to `all reviewers` |  | 
  `/wait` | Instruct Bert-E not to run until further notice. |  | 
  

</details>



<details>
  <summary><b>Available commands</b></summary>

  name   | description  | privileged
  ------ | ------------ | ------------
  `/help` | Print Bert-E's manual in the pull request. | 
  `/status` | Print Bert-E's current status in the pull request ```TBA``` | 
  `/clear` | Remove all comments from Bert-E from the history ```TBA``` | 
  `/retry` | Re-start a fresh build ```TBA``` | 
  `/build` | Re-start a fresh build ```TBA``` | 
  `/force_reset` | Delete integration branches & pull requests, and restart merge process from the beginning. | 
  `/reset` | Try to remove integration branches unless there are commits on them which do not appear on the source branch. | 
  
</details>




*Status report is not available.*

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue CLDSRV-638 contains:


* `9.0.10`



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `9.0.10`

* `9.1.0`


Please check the `Fix Version/s` of CLDSRV-638, or the target
branch of this pull request.

jonathan-gramain commented on tests/unit/Config.js:
```diff
@@ -1,8 +1,10 @@
     locationConstraintAssert,
     parseSupportedLifecycleRules,
-    ConfigObject: ConfigObjectForTest,
+    ConfigObject: ConfigObject,
```
```suggestion
    ConfigObject,
```

jonathan-gramain commented on lib/Config.js:
```diff
@@ -1770,6 +1766,12 @@ class Config extends EventEmitter {
         };
+
+        // Mongodb backend does not support null keys, so we must enforce null version compatibility
+        // mode. With other backends (esp. metadata), this is used during migration for v0 to v1
```
```suggestion
        // mode. With other backends (esp. metadata), this is used during migration from v0 to v1
```

jonathan-gramain commented on lib/api/objectHead.js:
```diff
@@ -151,11 +151,14 @@ function objectHead(authInfo, request, log, callback) {
                 const partSize = getPartSize(objMD, partNumber);
-                const isEmptyObject = objLength === 0;
-                if (!partSize && !isEmptyObject) {
+                if (partSize) {
```
You may want to squash this change with the previous commit's

jonathan-gramain commented on lib/metadata/acl.js:
```diff
@@ -41,8 +42,21 @@ const acl = {
+            // when multi-backend replication (i.e. Zenko) is used.
+            // TODO: this should be refactored to properly update the replication info, accounting
+            // for multiple rules and resetting the status only if needed CLDSRV-646
+            const replicationInfo = getReplicationInfo(objectKey, bucket, true);
```
Note for later: CLDSRV-647 is slated to add an extra parameter to this function call (to provide the S3 config explicitly), but it hasn't been reviewed yet so it may change.

francoisferrand committed:
Allow headObject on part1 for non mpu objects

Do not fail on head with part number if object is empty:
consider non-mpu objects as having 1 part, and add tests.

Issue: CLDSRV-638

bert-e commented:
# Request integration branches



Waiting for integration branch creation to be requested by the user.

To request integration branches, please comment on this pull request with the following command:

```
/create_integration_branches
```

Alternatively, the `/approve` and `/create_pull_requests` commands will automatically
create the integration branches.

williamlardier commented on eslint.config.mjs:
```diff
@@ -18,7 +18,7 @@ export default [...compat.extends('@scality/scality'), {
 
     languageOptions: {
-        ecmaVersion: 2020,
+        ecmaVersion: 2021,
```
Can we use the latest or some parts are not yet compatible (and should probably be identified)?

williamlardier reviewed (approved):
lgtm

francoisferrand commented:
/approve

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `bugfix/CLDSRV-638` into
`development/9.0`
* [w/9.1/bugfix/CLDSRV-638](https://github.com/scality/cloudserver/compare/development/9.1...w/9.1/bugfix/CLDSRV-638) will be merged into `development/9.1` 



The following branches will **NOT** be impacted:

* `development/7.10`
* `development/7.4`
* `development/7.70`
* `development/8.8`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```







*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/9.0`

* :heavy_check_mark: `development/9.1`



The following branches have **NOT** changed:

* `development/7.10`
* `development/7.4`
* `development/7.70`
* `development/8.8`



Please check the status of the associated issue CLDSRV-638.

Goodbye francoisferrand.




*The following options are set:* **approve**
